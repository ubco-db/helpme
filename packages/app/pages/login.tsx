import Router from 'next/router'
import { ReactElement, useEffect, useState } from 'react'
import { LeftOutlined, LockOutlined, UserOutlined } from '@ant-design/icons'
import { message, Button, Form, Input, Card, Select, Spin, Alert } from 'antd'
import styled from 'styled-components'
import Head from 'next/head'
import { API } from '@koh/api-client'
import useSWR from 'swr'
import ReCAPTCHA from 'react-google-recaptcha'
import React from 'react'

const Container = styled.div`
  margin-left: auto;
  margin-right: auto;
  text-align: center;
  padding-top: 100px;
  width: 60%;
  border-radius: 15px;
  height: auto;

  @media (max-width: 650px) {
    width: 90%;
  }

  @media (max-width: 992px) {
    width: 80%;
  }
`

export default function Login(): ReactElement {
  const [pass, setPass] = useState('')
  const [uname, setUname] = useState('')
  const [accountActiveResponse, setAccountActiveResponse] = useState(true)
  const [loginMenu, setLoginMenu] = useState(false)
  const [organization, setOrganization] = useState(null)

  const recaptchaRef = React.createRef()

  const { data: organizations } = useSWR(`api/v1/organization`, async () =>
    API.organizations.getOrganizations(),
  )

  const loginWithGoogle = async () => {
    await API.auth
      .loginWithGoogle(Number(organization.id))
      .then((res) => {
        Router.push(res.redirectUri)
      })
      .catch((err) => {
        console.log(err)
      })
  }

  const loginWithInstitution = async () => {
    Router.push(`/api/v1/auth/shibboleth/${organization.id}`)
  }

  async function login() {
    const token = await recaptchaRef.current.executeAsync()

    if (organization && !organization.legacyAuthEnabled) {
      message.error('Organization does not support legacy authentication')
      return
    }

    const loginRequest = {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        email: uname,
        password: pass,
        recaptchaToken: token,
      }),
    }
    fetch(`/api/v1/ubc_login`, loginRequest)
      .then(async (response) => {
        const data = await response.json()
        if (!response.ok) {
          // get error message from body or default to response statusText
          const error = (data && data.message) || response.statusText
          switch (response.status) {
            case 401:
              message.error(data.message)
              break
            case 403:
              setAccountActiveResponse(false)
              break
            case 404:
              message.error('User Not Found')
              break
            default:
              message.error(error)
              break
          }
          return Promise.reject(error)
        } else {
          Router.push(`/api/v1/login/entry?token=${data.token}`)
        }
      })
      .catch((error) => {
        console.error('There was an error!', error)
      })
  }

  const onPassChange = (e) => {
    setPass(e.target.value)
  }

  const onUserNameChange = (e) => {
    setUname(e.target.value)
  }

  const showLoginMenu = (value) => {
    const organization = organizations.find((org) => org.id === value)

    localStorage.setItem('organizationId', `${organization.id}`)

    if (!organization) {
      message.error('Organization not found')
      return
    }

    setOrganization(organization)
    setLoginMenu(true)
  }

  const onReCAPTCHAChange = (captchaCode) => {
    if (!captchaCode) {
      return
    }

    recaptchaRef.current.reset()
  }

  useEffect(() => {
    if (organizations && organizations.length === 1) {
      setOrganization(organizations[0])
      localStorage.setItem('organizationId', `${organizations[0].id}`)
    }
  }, [organizations, setLoginMenu])

  return organizations ? (
    <>
      <Head>
        <title>Login | HelpMe</title>
      </Head>
      <Container>
        <Card className="mx-auto max-w-md sm:px-2 md:px-6">
          <h2 className="my-4 text-left">Login</h2>

          {!loginMenu && (
            <>
              <p className="text-left text-stone-400">
                Select your organization.
              </p>

              <Select
                className="mt-2 w-full text-left"
                placeholder="Available Organizations"
                options={organizations.map((organization) => {
                  return {
                    label: organization.name,
                    value: organization.id,
                  }
                })}
                onChange={(value) => {
                  showLoginMenu(value)
                }}
              />
            </>
          )}

          {loginMenu && (
            <>
              {organizations && organizations.length > 1 && (
                <Button
                  className="flex w-full items-center justify-center gap-2 rounded-lg border px-5 py-5 text-left"
                  onClick={() => setLoginMenu(false)}
                >
                  <LeftOutlined />
                  <span className="font-semibold"> Go Back</span>
                </Button>
              )}

              {organization && organization.googleAuthEnabled && (
                <Button
                  className="mt-5 flex w-full items-center justify-center gap-2 rounded-lg border px-5 py-5 text-left"
                  onClick={() => loginWithGoogle()}
                >
                  <img
                    className="h-6 w-6"
                    src="https://www.svgrepo.com/show/475656/google-color.svg"
                    loading="lazy"
                    alt="google logo"
                  />
                  <span className="font-semibold">Log in with Google</span>
                </Button>
              )}

              {organization && organization.ssoEnabled && (
                <Button
                  className="mt-5 flex w-full items-center justify-center gap-2 rounded-lg border px-5 py-5 text-left"
                  onClick={() => loginWithInstitution()}
                >
                  <span className="font-semibold">Log in with Institution</span>
                </Button>
              )}

              {organization && organization.legacyAuthEnabled && (
                <p className="my-5 font-medium uppercase text-stone-400">
                  Or login with email
                </p>
              )}

              {!accountActiveResponse && (
                <Alert
                  message="System Notice"
                  description="Your account has been deactivated. Please contact your organization admin for more information."
                  type="error"
                  style={{ marginBottom: 20, textAlign: 'left' }}
                />
              )}
              {organization && organization.legacyAuthEnabled && (
                <Form
                  name="normal_login"
                  className="login-form"
                  initialValues={{ remember: true }}
                  onFinish={login}
                >
                  <ReCAPTCHA
                    ref={recaptchaRef}
                    size="invisible"
                    sitekey={process.env.NEXT_PUBLIC_RECAPTCHA_SITE_KEY}
                    onChange={onReCAPTCHAChange}
                  />
                  <Form.Item
                    name="username"
                    rules={[
                      {
                        required: true,
                        message: 'Please enter a valid username.',
                      },
                    ]}
                  >
                    <Input
                      prefix={<UserOutlined className="site-form-item-icon" />}
                      onChange={onUserNameChange}
                      className="rounded-lg border px-2 py-2"
                      placeholder="Username"
                    />
                  </Form.Item>

                  <Form.Item
                    name="password"
                    rules={[
                      {
                        required: true,
                        message: 'Please enter a valid password.',
                      },
                    ]}
                  >
                    <Input
                      prefix={<LockOutlined className="site-form-item-icon" />}
                      onChange={onPassChange}
                      type="password"
                      className="rounded-lg border px-2 py-2"
                      placeholder="Password"
                    />
                  </Form.Item>

                  <Form.Item>
                    <Button
                      type="primary"
                      htmlType="submit"
                      className="h-auto w-full items-center justify-center rounded-lg border px-2 py-2 "
                    >
                      <span className="font-semibold">Log in</span>
                    </Button>
                  </Form.Item>

                  <div className="d-flex flex-row space-x-8 text-center">
                    <a href="/account/password">Forgot password</a>
                    <a href="/register">Create account</a>
                  </div>
                </Form>
              )}
            </>
          )}
        </Card>
      </Container>
    </>
  ) : (
    <Spin />
  )
}
