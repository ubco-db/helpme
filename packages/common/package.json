{
  "name": "@koh/common",
  "version": "0.1.0",
  "main": "dist/index.js",
  "private": true,
  "scripts": {
    "tsc": "tsc"
  },
  "dependencies": {
    "class-transformer": "0.3.1",
    "class-validator": "^0.14.1",
    "reflect-metadata": "^0.1.13",
    "typescript": "^5.2.2"
  }
}
