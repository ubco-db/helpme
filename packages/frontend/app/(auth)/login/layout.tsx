import { cookies } from 'next/headers'
import { LoginRedirectInfoProvider } from './components/LoginRedirectInfoProvider'

export default async function Layout({
  children,
}: {
  children: React.ReactNode
}) {
  let invitedOrgId: number | null = null
  let invitedQueueId: number | null = null
  let invitedCourseId: number | null = null
  let invitedCourseInviteCode: string | null = null
  const cookieStore = await cookies()
  const queueInviteCookieString = cookieStore.get('queueInviteInfo')
  if (queueInviteCookieString) {
    const decodedCookie = decodeURIComponent(queueInviteCookieString.value)
    const cookieParts = decodedCookie.split(',')
    const courseId = cookieParts[0]
    const queueId = cookieParts[1]
    const orgId = cookieParts[2]
    const courseInviteCode = cookieParts[3]
      ? Buffer.from(cookieParts[3], 'base64').toString('utf-8')
      : null
    if (Number(orgId)) {
      invitedOrgId = Number(orgId)
    }
    if (Number(queueId)) {
      invitedQueueId = Number(queueId)
    }
    if (Number(courseId)) {
      invitedCourseId = Number(courseId)
    }
    if (courseInviteCode) {
      // note: this will only get set if the queueInvite had willInviteToCourse set to true
      invitedCourseInviteCode = courseInviteCode
    }
  } else {
    // get courseId from SECURE_REDIRECT (from invite code) and get the course's organization, and then set the organization to that
    const redirectCookieString = cookieStore.get('__SECURE_REDIRECT')
    if (redirectCookieString) {
      const decodedCookie = decodeURIComponent(redirectCookieString.value)
      const orgId = decodedCookie.split(',')[2]
      if (Number(orgId)) {
        invitedOrgId = Number(orgId)
      }
    }
  }
  return (
    <LoginRedirectInfoProvider
      invitedOrgId={invitedOrgId}
      invitedQueueId={invitedQueueId}
      invitedCourseId={invitedCourseId}
      invitedCourseInviteCode={invitedCourseInviteCode}
    >
      {children}
    </LoginRedirectInfoProvider>
  )
}
