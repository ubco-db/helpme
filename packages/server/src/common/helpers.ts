import { Request } from 'express';

export function getCookie(req: Request, cookieName: string): string {
  const cookieHeader = req.headers.cookie;
  if (!cookieHeader) {
    return null;
  }

  const cookies: string[] = cookieHeader.split(';');
  for (const cookie of cookies) {
    const [name, value] = cookie.split('=');
    if (name.trim() === cookieName) {
      return value;
    }
  }
}
