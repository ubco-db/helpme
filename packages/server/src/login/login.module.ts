import { Module } from '@nestjs/common';
import { LoginController } from './login.controller';
import { JwtStrategy } from '../login/jwt.strategy';
import { JwtModule } from '@nestjs/jwt';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { CourseService } from 'course/course.service';
import { RedisProfileModule } from 'redisProfile/redis-profile.module';
import { RedisProfileService } from 'redisProfile/redis-profile.service';
import { MailModule } from 'mail/mail.module';

@Module({
  imports: [
    JwtModule.registerAsync({
      imports: [ConfigModule, RedisProfileModule, MailModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) => ({
        secret: configService.get('JWT_SECRET'),
      }),
    }),
    MailModule,
  ],
  controllers: [LoginController],
  providers: [JwtStrategy, CourseService, RedisProfileService],
})
export class LoginModule {}
