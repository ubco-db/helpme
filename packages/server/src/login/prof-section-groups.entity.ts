import { KhouryProfCourse } from '@koh/common';
import { UserModel } from '../profile/user.entity';
import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  JoinColumn,
  OneToOne,
} from 'typeorm';

@Entity('prof_section_groups_model')
export class ProfSectionGroupsModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @OneToOne((type) => UserModel, {
    onDelete: 'CASCADE',
  })
  @JoinColumn({ name: 'profId' })
  prof: UserModel;

  @Column()
  profId: number;

  @Column({
    type: 'jsonb',
    nullable: true,
  })
  sectionGroups: KhouryProfCourse[];
}
