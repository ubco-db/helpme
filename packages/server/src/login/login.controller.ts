import { ERROR_MESSAGES, UBCOloginParam } from '@koh/common';
import {
  Body,
  Controller,
  Get,
  Req,
  HttpException,
  HttpStatus,
  Post,
  Query,
  Res,
  UnauthorizedException,
} from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { JwtService } from '@nestjs/jwt';
import { Request, Response } from 'express';
import * as bcrypt from 'bcrypt';
import { UserModel } from 'profile/user.entity';
import * as request from 'superagent';
import { getCookie } from 'common/helpers';

@Controller()
export class LoginController {
  constructor(
    private jwtService: JwtService,
    private configService: ConfigService,
  ) {}

  @Post('/ubc_login')
  async receiveDataFromLogin(
    @Res() res: Response,
    @Body() body: UBCOloginParam,
  ): Promise<any> {
    if (!body.recaptchaToken) {
      return res.status(HttpStatus.BAD_REQUEST).send({
        message: 'Recaptcha token missing',
      });
    }

    const response = await request.post(
      `https://www.google.com/recaptcha/api/siteverify?secret=${process.env.PRIVATE_RECAPTCHA_SITE_KEY}&response=${body.recaptchaToken}`,
    );

    if (!response.body.success) {
      return res.status(HttpStatus.BAD_REQUEST).send({
        message: 'Recaptcha token invalid',
      });
    }

    const user = await UserModel.findOne({
      where: { email: body.email },
      relations: ['organizationUser', 'organizationUser.organization'],
    });

    if (!user) {
      return res
        .status(HttpStatus.NOT_FOUND)
        .send({ message: 'User Not found' });
    }

    if (
      user.organizationUser &&
      user.organizationUser.organization.legacyAuthEnabled === false
    ) {
      return res.status(HttpStatus.UNAUTHORIZED).send({
        message: 'Organization does not allow legacy auth',
      });
    }

    const token = await this.jwtService.signAsync(
      { userId: user.id },
      { expiresIn: 60 },
    );

    if (token === null || token === undefined) {
      console.error('Temporary JWT is invalid');
      throw new HttpException(
        ERROR_MESSAGES.loginController.invalidTempJWTToken,
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }

    if (user.password === null || user.password === undefined) {
      return res.status(HttpStatus.UNAUTHORIZED).send({
        message: 'User did not sign up with legacy account system',
      });
    }

    bcrypt.compare(body.password, user.password, (err, data) => {
      //if error than throw error
      if (err) throw err;

      //if both match than you can do anything
      if (data) {
        if (user.accountDeactivated) {
          return res.status(HttpStatus.FORBIDDEN).send({
            message: 'Account deactivated',
          });
        }
        delete body.recaptchaToken;
        return res.status(200).send({ token, ...body });
      } else {
        return res.status(401).json({ message: 'Invalid credentials' });
      }
    });
  }

  // NOTE: Although the two routes below are on the backend,
  // they are meant to be visited by the browser so a cookie can be set

  // This is the real admin entry point, Kevin changed to also just take a user id, change to that sign in only
  @Get('/login/entry')
  async enterUBCOH(
    @Req() req: Request,
    @Res() res: Response,
    @Query('token') token: string,
  ): Promise<void> {
    const isVerified = await this.jwtService.verifyAsync(token);

    if (!isVerified) {
      throw new UnauthorizedException();
    }

    const payload = this.jwtService.decode(token) as { userId: number };
    await this.enter(req, res, payload.userId);
  }

  // Set cookie and redirect to proper page
  private async enter(req: Request, res: Response, userId: number) {
    // Expires in 30 days
    const authToken = await this.jwtService.signAsync({
      userId,
      expiresIn: 60 * 60 * 24 * 30,
    });

    if (authToken === null || authToken === undefined) {
      throw new HttpException(
        ERROR_MESSAGES.loginController.invalidTempJWTToken,
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }

    const cookie = getCookie(req, '__SECURE_REDIRECT');
    const redirectUrl = cookie
      ? Buffer.from(cookie, 'base64').toString('utf-8')
      : '/courses';

    const isSecure = this.configService
      .get<string>('DOMAIN')
      .startsWith('https://');
    res
      .cookie('auth_token', authToken, { httpOnly: true, secure: isSecure })
      .redirect(HttpStatus.FOUND, redirectUrl);
  }

  @Get('/logout')
  async logout(@Res() res: Response): Promise<void> {
    const isSecure = this.configService
      .get<string>('DOMAIN')
      .startsWith('https://');
    res
      .clearCookie('auth_token', { httpOnly: true, secure: isSecure })
      .redirect(302, '/login');
  }
}
