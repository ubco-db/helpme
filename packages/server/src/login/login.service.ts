import {
  HttpException,
  HttpStatus,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { ConfigService } from '@nestjs/config';
import { CourseService } from '../course/course.service';
import { LtiService } from '../lti/lti.service';
import { ERROR_MESSAGES } from '@koh/common';
import { CookieOptions, Request, Response } from 'express';
import { getCookie } from '../common/helpers';

export type LoginEntryOptions = {
  cookieName?: string;
  cookieOptions?: CookieOptions;
  restrictPaths?: (RegExp | string)[] | RegExp | string;
  expiresIn?: number;
  redirect?: string;
  returnImmediate?: boolean;
  returnImmediateMessage?: string;
};

@Injectable()
export class LoginService {
  constructor(
    private jwtService: JwtService,
    private configService: ConfigService,
  ) {}

  async initLoginEnter(
    req: Request,
    res: Response,
    token: string,
    courseService?: CourseService,
    ltiService?: LtiService,
    options: LoginEntryOptions = {},
  ) {
    const isVerified = await this.jwtService.verifyAsync(token);

    if (!isVerified) {
      throw new UnauthorizedException();
    }

    const payload = this.jwtService.decode(token) as { userId: number };
    await this.enter(
      req,
      res,
      payload.userId,
      courseService,
      ltiService,
      options,
    );
  }

  /**
   * @description Performs the entry mechanism to authorize users.
   * @param {Request} req Originating request
   * @param {Response} res Response object
   * @param {number} userId The UserId to be logged in
   * @param {CourseService} courseService The CourseService instance, if any, to use
   * @param {ltiService} ltiService The CourseService instance, if any, to use
   * @param {LoginEntryOptions} options Set of options to modify the generated token
   * @param {CookieOptions} options.cookieOptions Options to modify how the cookie is signed
   * @param {RegExp | string | (RegExp | string)[]} options.restrictPaths API/frontend paths this token should be allowed to use
   * @param {Number} options.expiresIn When the authorization token should expire
   * @param {String} options.redirect Override all other redirections to follow this path. Query parameters are extracted and applied after.
   * @Param {boolean} options.returnImmediate Override all redirections and cookies to return immediately with 200 OK.
   * @Param {String} options.returnImmediateMessage (Optional) Message to be sent with return immediate. Defaults to 'OK'.
   * @returns {void | Response} Returns 'void' if authorization is successful. Returns 500-level error response otherwise.
   */
  async enter(
    req: Request,
    res: Response,
    userId: number,
    courseService?: CourseService,
    ltiService?: LtiService,
    options: LoginEntryOptions = {},
  ): Promise<void | Response> {
    const {
      cookieName,
      restrictPaths,
      expiresIn,
      redirect,
      returnImmediate,
      returnImmediateMessage,
    } = options ?? {};
    let cookieOptions = options?.cookieOptions ?? { httpOnly: true };

    const secure =
      cookieOptions?.secure != undefined
        ? cookieOptions.secure
        : this.configService.get<string>('DOMAIN').startsWith('https://');

    cookieOptions = {
      ...cookieOptions,
      secure,
    };

    let authToken: string;
    try {
      authToken = await this.generateAuthToken(
        userId,
        expiresIn,
        restrictPaths,
      );
    } catch (err) {
      if (err instanceof HttpException) {
        return res.status(err.getStatus()).send(err.getResponse());
      } else {
        return res.status(500).send({ message: (err as Error).message });
      }
    }

    if (returnImmediate) {
      const ltiInvite = getCookie(req, '__COURSE_INVITE');
      if (ltiInvite && ltiService) {
        res = await this.handleLTICourseInviteCookie(
          res,
          userId,
          ltiInvite,
          ltiService,
        );
      }
      return res
        .cookie(cookieName ?? 'auth_token', authToken, cookieOptions)
        .send({ message: returnImmediateMessage ?? 'OK' });
    }

    let result = await this.handleCookies(
      req,
      res,
      userId,
      {
        cookieOptions,
        redirect,
        emailVerification: false,
      },
      courseService,
      ltiService,
    );

    if ('headersSent' in result && (result as Response).headersSent) {
      return;
    }

    result = result as { res: Response; redirectUrl: string };
    res = result.res;
    const { redirectUrl } = result;
    if (res.headersSent) {
      return;
    }

    return res
      .cookie(cookieName ?? 'auth_token', authToken, cookieOptions)
      .redirect(redirectUrl);
  }

  async handleCookies(
    req: Request,
    res: Response,
    userId: number,
    options: {
      cookieOptions: CookieOptions;
      redirect?: string;
      emailVerification?: boolean;
    },
    courseService?: CourseService,
    ltiService?: LtiService,
  ): Promise<Response | { res: Response; redirectUrl: string }> {
    const { emailVerification, cookieOptions } = options;
    let { redirect } = options;

    let redirectUrl: string;
    const initialParams: Record<string, string> = redirect
      ?.split('?')[1]
      ?.split('&')
      .map((v) => v.split('='))
      .reduce((p, c) => ({ ...p, [c[0]]: c[1] }), {});

    const paramIndex = redirect?.indexOf('?') ?? -1;
    if (paramIndex >= 0) {
      redirect = redirect.substring(0, paramIndex);
    }
    const queryParams = new URLSearchParams(initialParams);

    const secureRedirectCookie = getCookie(req, '__SECURE_REDIRECT');
    const queueInviteCookie = getCookie(req, 'queueInviteInfo');
    const ltiInviteCookie = getCookie(req, '__COURSE_INVITE');

    if (queueInviteCookie && courseService) {
      await courseService
        .getQueueInviteRedirectURLandInviteToCourse(queueInviteCookie, userId)
        .then((result) => {
          redirectUrl = result.url;
          for (const [key, value] of result.queryParams) {
            queryParams.set(key, value);
          }
          res.clearCookie('queueInviteInfo', cookieOptions);
        });
    } else if (secureRedirectCookie) {
      const decodedCookie = decodeURIComponent(secureRedirectCookie);
      const cookieParts = decodedCookie.split(',');
      queryParams.set('cid', cookieParts[0]);
      queryParams.set('code', encodeURIComponent(cookieParts[1]));
      redirectUrl = `/invite`;
      res.clearCookie('__SECURE_REDIRECT', cookieOptions);
    } else if (ltiInviteCookie && ltiService) {
      const inviteToken = decodeURIComponent(ltiInviteCookie);
      let courseId: number | undefined;
      if (inviteToken) {
        try {
          courseId = await ltiService.checkCourseInvite(userId, inviteToken);
        } catch (err) {
          queryParams.set('err', (err as Error).message);
        }
      }
      if (courseId) {
        redirectUrl = `/lti/${courseId}`;
      } else {
        redirectUrl = `/lti`;
      }
      res.clearCookie('__COURSE_INVITE', cookieOptions);
    } else if (redirect) {
      redirectUrl = redirect;
    } else {
      redirectUrl = emailVerification ? undefined : '/courses';
    }

    redirectUrl =
      redirectUrl != undefined
        ? `${redirectUrl}${queryParams.size > 0 ? '?' + queryParams.toString() : ''}`
        : undefined;

    if (emailVerification && redirectUrl) {
      return res.status(HttpStatus.TEMPORARY_REDIRECT).send({
        redirectUri: redirectUrl,
      });
    } else if (emailVerification) {
      return res.status(HttpStatus.ACCEPTED).send({
        message: 'Email verified',
      });
    }

    return {
      res,
      redirectUrl,
    };
  }

  async handleLTICourseInviteCookie(
    res: Response,
    userId: number,
    inviteCookie: string,
    ltiService: LtiService,
  ) {
    const inviteToken = decodeURIComponent(inviteCookie);
    if (inviteToken) {
      try {
        await ltiService.checkCourseInvite(userId, inviteToken);
        res.clearCookie('__COURSE_INVITE', LtiService.cookieOptions);
      } catch (err) {
        console.error(`Error occurred handling LTI course invite`, err);
      }
    }
    return res;
  }

  async generateAuthToken(
    userId: number,
    expiresIn: number = 60 * 60 * 24 * 30, // Expires in 30 days (Default)
    restrictPaths?: (RegExp | string) | (RegExp | string)[],
  ) {
    const authToken = await this.jwtService.signAsync({
      userId,
      expiresIn,
      restrictPaths,
    });

    if (authToken === null || authToken === undefined) {
      throw new HttpException(
        ERROR_MESSAGES.loginController.invalidTempJWTToken,
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }

    return authToken;
  }
}
