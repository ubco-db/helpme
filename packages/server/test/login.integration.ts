import { JwtService } from '@nestjs/jwt';
import { LoginModule } from '../src/login/login.module';
import { OrganizationFactory, UserFactory } from './util/factories';
import { setupIntegrationTest } from './util/testUtils';
import * as bcrypt from 'bcrypt';
import { OrganizationUserModel } from 'organization/organization-user.entity';
import '@koh/common';

jest.mock('@koh/common', () => ({
  ...jest.requireActual('@koh/common'),
  isProd: () => true,
}));

jest.mock('superagent', () => ({
  post: jest.fn().mockImplementation((url) => {
    if (url.includes('invalid')) {
      return { body: { success: false } };
    }
    return { body: { success: true } };
  }),
}));

describe('Login Integration', () => {
  let jwtService: JwtService;

  const { supertest, getTestModule } = setupIntegrationTest(LoginModule);

  beforeAll(() => {
    jwtService = getTestModule().get<JwtService>(JwtService);
  });

  describe('POST /login', () => {
    it('returns 400 if no email is provided', async () => {
      await supertest()
        .post('/login')
        .send({ password: 'fake_password', recaptchaToken: 'token' })
        .expect(400);
    });

    it('returns 400 if recaptcha returned response false', async () => {
      await supertest()
        .post('/login')
        .send({
          email: 'fake_email@ubc.ca',
          password: 'fake_password',
          recaptchaToken: 'invalid',
        })
        .expect(400);
    });

    it('returns 404 if user not found', async () => {
      await supertest()
        .post('/login')
        .send({
          email: 'fake_email@ubc.ca',
          password: 'fake_password',
          recaptchaToken: 'token',
        })
        .expect(404);
    });

    it('returns 401 if password is incorrect', async () => {
      const user = await UserFactory.create({ password: 'real_password' });

      await supertest()
        .post('/login')
        .send({
          email: user.email,
          password: 'invalid_password',
          recaptchaToken: 'token',
        })
        .expect(401);
    });

    it('returns 403 if account is deactivated', async () => {
      const salt = await bcrypt.genSalt(10);
      const password = await bcrypt.hash('realpassword', salt);

      const user = await UserFactory.create({
        password: password,
        accountDeactivated: true,
      });

      await supertest()
        .post('/login')
        .send({
          email: user.email,
          password: 'realpassword',
          recaptchaToken: 'token',
        })
        .expect(403);
    });

    it('returns 200 if password is correct', async () => {
      const salt = await bcrypt.genSalt(10);
      const password = await bcrypt.hash('realpassword', salt);

      const user = await UserFactory.create({ password: password });

      const res = await supertest().post('/login').send({
        email: user.email,
        password: 'realpassword',
        recaptchaToken: 'token',
      });

      const token = '' + res.body.token;
      const decoded = jwtService.decode(token);
      expect(decoded).toEqual({
        userId: user.id,
        exp: expect.any(Number),
        iat: expect.any(Number),
      });
      delete res.body.token;
      expect(res.body).toMatchSnapshot();
      expect(res.status).toBe(200);
    });

    it('returns 405 when organization does not allow legacy auth', async () => {
      const organization = await OrganizationFactory.create({
        legacyAuthEnabled: false,
      });
      const user = await UserFactory.create({ password: 'real_password' });

      await OrganizationUserModel.create({
        userId: user.id,
        organizationId: organization.id,
      }).save();

      await supertest()
        .post('/login')
        .send({
          email: user.email,
          password: 'real_password',
          recaptchaToken: 'token',
        })
        .expect(405);
    });

    it('returns 418 when user did not sign up with legacy account system', async () => {
      const organization = await OrganizationFactory.create({
        legacyAuthEnabled: true,
      });
      const user = await UserFactory.create({ password: null });

      await OrganizationUserModel.create({
        userId: user.id,
        organizationId: organization.id,
      }).save();

      await jwtService.signAsync({ userId: user.id });

      await supertest()
        .post('/login')
        .send({
          email: user.email,
          password: 'real_password',
          recaptchaToken: 'token',
        })
        .expect(418);
    });
  });

  describe('POST /login/entry', () => {
    it('request to entry with correct jwt payload works', async () => {
      const user = await UserFactory.create();
      const token = await jwtService.signAsync({ userId: user.id });

      const res = await supertest()
        .get(`/login/entry?token=${token}`)
        .expect(302);

      expect(res.header['location']).toBe('/courses');

      const cookie = res.get('Set-Cookie')[0];
      const mainPart = cookie.substring(0, cookie.indexOf(';'));
      const secondPart = cookie.substring(cookie.indexOf(';') + 1);
      const [name, value] = mainPart.split('=');

      expect(name).toEqual('auth_token');

      const jwtToken = jwtService.decode(value);

      expect(jwtToken).toEqual(
        expect.objectContaining({
          userId: user.id,
          expiresIn: 24 * 30 * 60 * 60,
          iat: expect.anything(),
        }),
      );

      const parts = secondPart.split(';').map((v) => v.trim());
      const flags = parts.slice(1);

      expect(flags).toHaveLength(1);
      expect(flags[0]).toBe('HttpOnly');
    });

    it('request to entry with invalid jwt returns error', async () => {
      const user = await UserFactory.create();
      const token = await jwtService.signAsync({ userId: user.id });

      const spy = jest.spyOn(JwtService.prototype, 'verifyAsync');
      spy.mockResolvedValue(null);

      await supertest().get(`/login/entry?token=${token}`).expect(401);

      spy.mockRestore();
    });
  });

  describe('GET /logout', () => {
    it('makes sure logout endpoint is destroying cookies like a mob boss', async () => {
      const res = await supertest().get(`/logout`).expect(302);
      expect(res.header['location']).toBe('/login');
      expect(res.get('Set-Cookie')[0]).toContain('auth_token=;');
    });
  });
});
