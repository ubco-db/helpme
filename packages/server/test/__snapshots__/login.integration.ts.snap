// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Login Integration POST /ubc_login returns 200 if password is correct 1`] = `
{
  "email": "user@ubc.ca",
  "password": "realpassword",
  "token": "{"userId":1}",
}
`;
